q2_feature_table/_heatmap/_visualizer.py

```diff
@@ -0,0 +1,151 @@
+                     'Spectral_r', 'Vega10', 'Vega10_r', 'Vega20', 'Vega20_r',
+                     'Vega20b', 'Vega20b_r', 'Vega20c', 'Vega20c_r', 'Wistia',
+                     'Wistia_r', 'YlGn', 'YlGnBu', 'YlGnBu_r', 'YlGn_r',
+                     'YlOrBr', 'YlOrBr_r', 'YlOrRd', 'YlOrRd_r', 'afmhot',
+                     'afmhot_r', 'autumn', 'autumn_r', 'binary', 'binary_r',
+                     'bone', 'bone_r', 'brg', 'brg_r', 'bwr', 'bwr_r', 'cool',
+                     'cool_r', 'coolwarm', 'coolwarm_r', 'copper', 'copper_r',
+                     'cubehelix', 'cubehelix_r', 'flag', 'flag_r',
+                     'gist_earth', 'gist_earth_r', 'gist_gray', 'gist_gray_r',
+                     'gist_heat', 'gist_heat_r', 'gist_ncar', 'gist_ncar_r',
+                     'gist_rainbow', 'gist_rainbow_r', 'gist_stern',
+                     'gist_stern_r', 'gist_yarg', 'gist_yarg_r', 'gnuplot',
+                     'gnuplot2', 'gnuplot2_r', 'gnuplot_r', 'gray', 'gray_r',
+                     'hot', 'hot_r', 'hsv', 'hsv_r', 'icefire', 'icefire_r',
+                     'inferno', 'inferno_r', 'jet', 'jet_r', 'magma',
+                     'magma_r', 'mako', 'mako_r', 'nipy_spectral',
+                     'nipy_spectral_r', 'ocean', 'ocean_r', 'pink', 'pink_r',
+                     'plasma', 'plasma_r', 'prism', 'prism_r', 'rainbow',
+                     'rainbow_r', 'rocket', 'rocket_r', 'seismic', 'seismic_r',
+                     'spectral', 'spectral_r', 'spring', 'spring_r', 'summer',
+                     'summer_r', 'tab10', 'tab10_r', 'tab20', 'tab20_r',
+                     'tab20b', 'tab20b_r', 'tab20c', 'tab20c_r', 'terrain',
+                     'terrain_r', 'viridis', 'viridis_r', 'vlag', 'vlag_r',
+                     'winter', 'winter_r'}
+}
+
+
+def _normalizer(x):
+    return np.log(x + 1)
+
+
+def heatmap(output_dir, table: pd.DataFrame,
+            metadata: qiime2.MetadataCategory=None, normalize: bool=True,
+            metric: str='euclidean', method: str='average',
+            cluster: str='both', color_scheme: str='rocket') -> None:
+    clustering_map = {'both': {'col_cluster': True, 'row_cluster': True},
+                      'samples': {'col_cluster': False, 'row_cluster': True},
+                      'features': {'col_cluster': True, 'row_cluster': False}}
+
+    category = metadata.to_series() if metadata is not None else None
+
+    # Validation
+    if table.empty:
+        raise ValueError('Provided table is empty.')
+    if category is not None:
+        metadata_ids = set(category.index.tolist())
+        table_ids = set(table.index.tolist())
+        if not table_ids.issubset(metadata_ids):
+            raise ValueError('Missing samples in metadata: %r' %
+                             table_ids.difference(metadata_ids))
+
+    cbar_label = 'frequency'
+    if normalize:
+        table = table.apply(_normalizer)
+        cbar_label = 'log10 frequency'
+
+    if category is not None:
```

Comment:
I should probably break this whole block out into a testable util function, it is doing a lot of untested jazz. 🎷